{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Deliverables:  36 total points\n",
    "\n",
    " - seqCleaner  9 points\n",
    " - fastqParse  9 points\n",
    " - coordinateMathSoln 9 points\n",
    " - converter 9 points\n",
    " - Optional: Triad class rewrite (5 points extra credit)\n",
    " \n",
    " These must be submitted as 4(or 5) .py files, \n",
    "     \n",
    " Due: Monday April 12, 2021 11:55pm\n",
    "\n",
    "# Lab 2 â€“ Manipulating Data Types (36 points)\n",
    "## Overview\n",
    "Last week we introduced the Python function input(), which is used to take in string data. This week we will use input() to take in string and numeric (float and integer) data. \n",
    "\n",
    "Remember that all data returned from input() is a string object, so this will mean that you need to convert any numeric data to their respective numeric objects. The exercises in this lab will give you practice manipulating various types of data that commonly arise in computational biology problems.\n",
    "\n",
    "We are again using jupyter for this assignment, and as a final step we will create distinct command line programs from each of the parts.\n",
    "Please save each of the four (or five) python programs as .py files into your LAB02 folder. You can copy the text from the code cell, then paste it into an editor like notepad or textedit. Save that new file with the appropriate name ( eg. seqCleaner.py) into your LAB02 folder. Using terminal or cmd, you can now do a final test on that new program with the command:<br>\n",
    "python3 seqCleaner <br>\n",
    "\n",
    "All together, you will submit:\n",
    " - seqCleaner.py, \n",
    " - fastqParse,py, \n",
    " - coordinateMathSoln.py, \n",
    " - converter.py, and \n",
    " - (optionally) Triad.py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### seqCleaner"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DNA data?AAttCCCaaaTTNNNNNNGGggCCcccaAAAAATTTTTTtttt\n",
      "Clean Sequence:\n",
      "AATTCCCAAATT{6}GGGGCCCCCAAAAAATTTTTTTTTT\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3 \n",
    "# Name: Grace Jacobson  \n",
    "# Group Members: None\n",
    "\n",
    "\"\"\"\"\n",
    "seqCleaner\n",
    "\n",
    "input: a DNA sequence in upper or lower case with AT MOST one block of ambiguous N characters.\n",
    "output: a DNA sequence in upper case with a block of N characters replaces with the count of ambiguous bases in brackets.\n",
    "\n",
    "A regular expression is used to replace one or more -indicated by + sign-  Ns in the sequence\n",
    "\n",
    "Sources\n",
    "https://docs.python.org/3/library/re.html \n",
    "https://cheatography.com/davechild/cheat-sheets/regular-expressions/ \n",
    "\n",
    "\"\"\"\"\n",
    "import re #regular expression\n",
    "\n",
    "class DNAstring (str):\n",
    "    \"\"\"This class represents the DNA string input that needs to be cleaned. \n",
    "    Contains a purify method to replace a block of Ns.\n",
    "    \"\"\"\n",
    "    def length (self):\n",
    "        return (length(self))\n",
    "    \n",
    "    def purify(self):\n",
    "        ''' Return an upcased version of the string, collapsing a single run of Ns.'''\n",
    "        numberN = self.count(\"N\") #counts the number of N in a single block\n",
    "        return  re.sub('N+','{' + str(numberN) + '}', self) #replaces a block of one or more Ns with the number of N counted, within brackets\n",
    "        \n",
    "def main():\n",
    "    ''' Get user DNA data and clean it up.'''\n",
    "    data = input('DNA data?').upper() #make all bases uppercase in input\n",
    "    thisDNA = DNAstring (data) #creates object of DNAstring class with input\n",
    "    pureData = thisDNA.purify() #applies purify method to data\n",
    "    print (pureData) #output\n",
    "    \n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### fastqParse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter FASTQ seqname line:@EAS139:136:FC706VJ:2:2104:15343:197393\n",
      "Instrument = EAS139\n",
      "Run ID = 136\n",
      "Flow Cell ID = FC706VJ\n",
      "Flow Cell Lane = 2\n",
      "Tile Number = 2104\n",
      "X-coord = 15343\n",
      "Y-coord = 197393\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3 \n",
    "# Name: Grace Jacobson\n",
    "# Group Members: None\n",
    "\n",
    "'''\n",
    "fastqParse\n",
    "\n",
    "input: a FASTQ seqname line starting with @, with seven fields separated by :\n",
    "output: run information with each field on a new line, with an = sign\n",
    "\n",
    "\n",
    "'''\n",
    "\n",
    "class FastqString (str):\n",
    "    ''' A class to represent a FASTQ string containing run information. Contains parse method to separate each field.'''\n",
    "    \n",
    "    def parse(self):\n",
    "        ''' This method will separate each field and print the run information.'''\n",
    "        parseList = self.split(\":\") #forms a list of each field, which were separated by : . the first item is a blank space therefore the program starts printing at item 1\n",
    "        print(\"Instrument = \"+parseList[1]) #prints the second item in the list, the instrument\n",
    "        print(\"Run ID = \"+parseList[2]) #prints the third item in the list, the Run ID\n",
    "        print(\"Flow Cell ID = \"+parseList[3]) #prints the fourth item in the list, the flow cell ID\n",
    "        print(\"Flow Cell Lane = \"+parseList[4]) #prints the fifth item in the list, the flwo cell lane\n",
    "        print(\"Tile Number = \"+parseList[5]) #prints the sixth item in the list, the tile number\n",
    "        print(\"X-coord = \"+parseList[6]) #prints the seventh item in the list, the x-coordinate\n",
    "        print(\"Y-coord = \"+parseList[7]) #prints the eighth item in the list, the y-coordinate\n",
    "    \n",
    "def main():\n",
    "    ''' The main function will take an FASTQ seqname line and separate it from the @ symbol and direct to the parse method '''\n",
    "    data = input(\"Enter FASTQ seqname line:\") \n",
    "    fastqList = data.split('@') #separates the seven fields from the @ symbol, creating a list with two items of an empty space and the rest of the seqname line\n",
    "    fastqString = \":\".join(fastqList) #creates a string, joining the space to the rest of the seven fields with a :\n",
    "    FastqString.parse(fastqString) #applies the parse method of the FastqString object to the string\n",
    "    \n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### coordinateMathSoln"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter Coordinates: C = (39.447, 94.657, 11.824) N = (39.292, 95.716, 11.027) Ca = (39.462, 97.101, 11.465)\n",
      "N-C bond length = 1.33\n",
      "N-Ca bond length = 1.46\n",
      "C-N-Ca bond angle = 124.0\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3 \n",
    "# Name: Grace Jacobson\n",
    "# Group Members: None\n",
    "\n",
    "'''\n",
    "coordinateMathSoln\n",
    "\n",
    "Takes coordinates contained in parantheses for C, N, and Ca atoms with an '=' sign in one input line. \n",
    "Prints N-C and N-Ca bond lengths and C-N-Ca bond angle in degrees as its output.\n",
    "\n",
    "'''\n",
    "\n",
    "import math\n",
    "import re\n",
    "\n",
    "class Triad :\n",
    "    \"\"\"\n",
    "    Calculate angles and distances among a triad of points.\n",
    " \n",
    "    Author: David Bernick\n",
    "    Date: March 21, 2013\n",
    "    Points can be supplied in any dimensional space as long as they are consistent.\n",
    "    Points are supplied as tupels in n-dimensions, and there should be three\n",
    "    of those to make the triad. Each point is positionally named as p,q,r\n",
    "    and the corresponding angles are then angleP, angleQ and angleR.\n",
    "    Distances are given by dPQ(), dPR() and dQR()\n",
    " \n",
    "    Required Modules: math\n",
    "    initialized: 3 positional tuples representing Points in n-space\n",
    "             p1 = Triad( p=(1,0,0), q=(0,0,0), r=(0,1,0) )\n",
    "    attributes: p,q,r the 3 tuples representing points in N-space\n",
    "    methods:  angleP(), angleR(), angleQ() angles measured in radians\n",
    "          dPQ(), dPR(), dQR() distances in the same units of p,q,r\n",
    " \n",
    "    \"\"\"\n",
    " \n",
    "    def __init__(self,p,q,r) :\n",
    "        \"\"\" Construct a Triad. \n",
    "        Example construction:\n",
    "            p1 = Triad( p=(1.,0.,0.), q=(0.,0.,0.), r=(0.,0.,0.) ). \n",
    "        \"\"\"\n",
    "        self.p = p\n",
    "        self.q = q\n",
    "        self.r = r\n",
    "        \n",
    "# private helper methods\n",
    "    def d2 (self,a,b) : # calculate squared distance of point a to b\n",
    "        return float(sum((ia-ib)*(ia-ib)  for  ia,ib in zip (a,b)))\n",
    "    \n",
    "    def dot (self,a,b) : # dotProd of standard vectors a,b\n",
    "        return float(sum(ia*ib for ia,ib in zip(a,b)))\n",
    "    \n",
    "    def ndot (self,a,b,c) : # dotProd of vec. a,c standardized to b\n",
    "        return float(sum((ia-ib)*(ic-ib) for ia,ib,ic in zip (a,b,c)))\n",
    "    \n",
    "# calculate lengths(distances) of segments PQ, PR and QR\n",
    "    def dPQ (self):\n",
    "        \"\"\" Provides the distance between point p and point q \"\"\"\n",
    "        return math.sqrt(self.d2(self.p,self.q))\n",
    "    \n",
    "    def dPR (self):\n",
    "        \"\"\" Provides the distance between point p and point r \"\"\"\n",
    "        return math.sqrt(self.d2(self.p,self.r))\n",
    "    \n",
    "    def dQR (self):\n",
    "        \"\"\" Provides the distance between point q and point r \"\"\"\n",
    "        return math.sqrt(self.d2(self.q,self.r))\n",
    "    \n",
    "    def angleP (self) :\n",
    "        \"\"\" Provides the angle made at point p by segments pq and pr (radians). \"\"\"\n",
    "        return math.acos(self.ndot(self.q,self.p,self.r) /   math.sqrt(self.d2(self.q,self.p)*self.d2(self.r,self.p)))\n",
    "    \n",
    "    def angleQ (self) :\n",
    "        \"\"\" Provides the angle made at point q by segments qp and qr (radians). \"\"\"\n",
    "        return math.acos(self.ndot(self.p,self.q,self.r) /  math.sqrt(self.d2(self.p,self.q)*self.d2(self.r,self.q)))\n",
    " \n",
    "    def angleR (self) :\n",
    "        \"\"\" Provides the angle made at point r by segments rp and rq (radians). \"\"\"\n",
    "        return math.acos(self.ndot(self.p,self.r,self.q) /  math.sqrt(self.d2(self.p,self.r)*self.d2(self.q,self.r)))\n",
    "\n",
    "def main():\n",
    "    ''' \n",
    "    This function uses regular expressions to find three coordinates for C, N, and Ca. Using a regular\n",
    "    expression allows the program to find the coordinates for each point no matter the number of spaces between\n",
    "    each atom and '=' sign. After forming the triad, the function prints each bond length and bond angle. \n",
    "    The bond angle is converted to degrees.\n",
    "    \n",
    "    Resources:\n",
    "            https://www.regular-expressions.info/floatingpoint.html for regular expressions to find coordinates\n",
    "            https://regexr.com/ for editing the regular expression\n",
    "            https://www.geeksforgeeks.org/python-convert-list-of-tuples-into-list/ for converting tuples into a list\n",
    "            \n",
    "    '''\n",
    "    coordinates = input(\"Enter Coordinates: \")\n",
    "    \n",
    "    #creating list of p coordinates using a regular expression\n",
    "    pTuples = re.findall(\".*C+\\s*=\\s*\\([-+]?([0-9]*\\.[0-9]+|[0-9]+)\\s*,\\s*[-+]?([0-9]*\\.[0-9]+|[0-9]+)\\s*,\\s*[-+]?([0-9]*\\.[0-9]+|[0-9]+)\", coordinates)\n",
    "    p = [] #creates p as a list of coordinates\n",
    "    for x in pTuples: #converts the tuple into a mutable list\n",
    "            for t in x:\n",
    "                p.append(float(t))\n",
    "                \n",
    "    #creating a list of q coordinates using a regular expression\n",
    "    qTuples = re.findall(\".*N+\\s*=\\s*\\([-+]?([0-9]*\\.[0-9]+|[0-9]+)\\s*,\\s*[-+]?([0-9]*\\.[0-9]+|[0-9]+)\\s*,\\s*[-+]?([0-9]*\\.[0-9]+|[0-9]+)\", coordinates)\n",
    "    q = [] #creates q as a list of coordinates\n",
    "    for x in qTuples: #converts the tuple into a mutable list\n",
    "            for t in x:\n",
    "                q.append(float(t))\n",
    "    \n",
    "    #creating a list of r coordinates using a regular expression\n",
    "    rTuples = re.findall(\".*Ca+\\s*=\\s*\\([-+]?([0-9]*\\.[0-9]+|[0-9]+)\\s*,\\s*[-+]?([0-9]*\\.[0-9]+|[0-9]+)\\s*,\\s*[-+]?([0-9]*\\.[0-9]+|[0-9]+)\", coordinates)\n",
    "    r = [] #creates r as a list of coordinates\n",
    "    for x in rTuples: #converts the tuple into a mutable list\n",
    "            for t in x:\n",
    "                r.append(float(t))\n",
    "    \n",
    "    p1 = Triad(p,q,r) #forms triad\n",
    "    print(\"N-C bond length = {0:0.2f}\".format(p1.dPQ())) #prints the bond length with 2 decimal places\n",
    "    print(\"N-Ca bond length = {0:0.2f}\".format(p1.dQR())) #prints the bond length with 2 decimal places\n",
    "    \n",
    "    angleDegrees = math.degrees(p1.angleQ()) #i used math.degrees to convert angleQ to degrees\n",
    "    print(\"C-N-Ca bond angle = {0:0.1f}\".format(angleDegrees)) #prints the bond angle in degrees with one decimal place\n",
    "    \n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### converter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter RNA or DNA codon, or amino acid(1 letter or 3 letter code):TAT\n",
      "TAT = Tyr\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3 \n",
    "# Name: Grace Jacobson\n",
    "# Group Members: None\n",
    "\n",
    "'''\n",
    "converter  input \n",
    "\n",
    "This program can:\n",
    "Take short amino acid abbreviation input and convert to long amino acid abbreviation.\n",
    "OR\n",
    "Take long amino acid abbreviation input and convert to short amino acid abbreviation.\n",
    "OR\n",
    "Take RNA or DNA three letter codon input and convert to long amino acid abbreviation.\n",
    "'''\n",
    "\n",
    "short_AA = {\n",
    "            'CYS': 'C', 'ASP': 'D', 'SER': 'S', 'GLN': 'Q', 'LYS': 'K',\n",
    "            'ILE': 'I', 'PRO': 'P', 'THR': 'T', 'PHE': 'F', 'ASN': 'N', \n",
    "            'GLY': 'G', 'HIS': 'H', 'LEU': 'L', 'ARG': 'R', 'TRP': 'W', \n",
    "            'ALA': 'A', 'VAL': 'V', 'GLU': 'E', 'TYR': 'Y', 'MET': 'M'\n",
    "            }\n",
    "\n",
    "long_AA = {value:key for key,value in short_AA.items()}\n",
    "\n",
    "RNA_codon_table = {\n",
    "                    # Second Base\n",
    "# U             C             A             G\n",
    "#U\n",
    "'UUU': 'Phe', 'UCU': 'Ser', 'UAU': 'Tyr', 'UGU': 'Cys',\n",
    "'UUC': 'Phe', 'UCC': 'Ser', 'UAC': 'Tyr', 'UGC': 'Cys',\n",
    "'UUA': 'Leu', 'UCA': 'Ser', 'UAA': '---', 'UGA': '---',\n",
    "'UUG': 'Leu', 'UCG': 'Ser', 'UAG': '---', 'UGG': 'Trp',\n",
    "#C \n",
    "'CUU': 'Leu', 'CCU': 'Pro', 'CAU': 'His', 'CGU': 'Arg',\n",
    "'CUC': 'Leu', 'CCC': 'Pro', 'CAC': 'His', 'CGC': 'Arg',\n",
    "'CUA': 'Leu', 'CCA': 'Pro', 'CAA': 'Gln', 'CGA': 'Arg',\n",
    "'CUG': 'Leu', 'CCG': 'Pro', 'CAG': 'Gln', 'CGG': 'Arg',\n",
    "#A\n",
    "'AUU': 'Ile', 'ACU': 'Thr', 'AAU': 'Asn', 'AGU': 'Ser',\n",
    "'AUC': 'Ile', 'ACC': 'Thr', 'AAC': 'Asn', 'AGC': 'Ser',\n",
    "'AUA': 'Ile', 'ACA': 'Thr', 'AAA': 'Lys', 'AGA': 'Arg',\n",
    "'AUG': 'Met', 'ACG': 'Thr', 'AAG': 'Lys', 'AGG': 'Arg',\n",
    "#G\n",
    "'GUU': 'Val', 'GCU': 'Ala', 'GAU': 'Asp', 'GGU': 'Gly',\n",
    "'GUC': 'Val', 'GCC': 'Ala', 'GAC': 'Asp', 'GGC': 'Gly',\n",
    "'GUA': 'Val', 'GCA': 'Ala', 'GAA': 'Glu', 'GGA': 'Gly',\n",
    "'GUG': 'Val', 'GCG': 'Ala', 'GAG': 'Glu', 'GGG': 'Gly'\n",
    "}\n",
    "dna_Codon_Table = {key.replace('U','T'):value for key, value in RNA_codon_table.items()}\n",
    "\n",
    "def main():\n",
    "\"\"\"The main function takes the input data and prints the value for the dictionary in which it is a key.\n",
    "    If it does not match a key in any dictionary, it will print unknown.\"\"\"\"\n",
    "    data = input('Enter RNA codon, DNA codon, or amino acid(1 letter or 3 letter code):')\n",
    "    key = data.upper() #makes the input uppercase to match keys in all dictionaries, in case input is lowercase\n",
    "    \n",
    "    if key in short_AA.keys(): #identifies the key in short amino acid abbreviation dictionary\n",
    "        answerShortAA = short_AA.get(key) #the answer is the value for the key \n",
    "        print(key + \" = \" + answerShortAA) #prints the input with equal to the converted amino acid abbreviation\n",
    "        \n",
    "    if key in long_AA.keys(): #identifies key in the long amino acid abbreviation dictionary\n",
    "        answerLongAA = long_AA.get(key) #the answer is the value for the key in the dictionary\n",
    "        print(key + \" = \" + answerLongAA)#prints the input equal to the converted amino acid abbreviation\n",
    "        \n",
    "    if key in dna_Codon_Table.keys(): #identifies the key in the DNA codon table\n",
    "        answerDNA = dna_Codon_Table.get(key) #answer is the value for the key in the dictionary\n",
    "        print(key + \" = \" + answerDNA) #prints the input equal to the converted amino acid long abbreviation \n",
    "        \n",
    "    if key in RNA_codon_table.keys(): #identifies the input as a key in the RNA codon table\n",
    "        answerRNA = RNA_codon_table.get(key) #answer is the value for the key in the dictionary\n",
    "        print(key + \" = \" + answerRNA) #prints input equal to the converted amino acid long abbreviation\n",
    "    \n",
    "    #if the input is not a key in any of the dictionaries, it will print unknown\n",
    "    if (key not in short_AA.keys() and key not in long_AA.keys() and key not in dna_Codon_Table.keys() and key not in RNA_codon_table.keys()):\n",
    "        print(\"Unknown\")\n",
    "       \n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Submit your code and answers\n",
    " \n",
    "Important: please save your work before logging out of the computer. This will ensure that you have a copy of your work and you will avoid having to redo everything in the  event that something happens with the lab computers. The two recommended solutions (pick one) at this point are to:\n",
    " - Email your code files to yourself\n",
    " - Copy your code files to a pen drive\n",
    " - Save your work on your own computer if you are using Anaconda ( or equiv)\n",
    "\n",
    "The class will be using canvas to submit the source code files created as part of this lab assignment.\n",
    "For this lab, you should upload the following programs as files:\n",
    " \n",
    " - seqCleaner.py 9 points\n",
    " - fastqParse.py 9 points\n",
    " - coordinateMathSoln.py 9 points\n",
    " - converter.py 9 points\n",
    " - triad.py Triad rewrite (optional) 5 points\n",
    " \n",
    "Important: to get full credit on this lab assignment, each of the code files you submit needs to:\n",
    " - Run properly (execute and produce the correct output)\n",
    " - Contain proper docstrings and appropriate line style comments (#)\n",
    " \n",
    "Congratulations, you finished your second lab assignment!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
